File: examples/typescript/order-repository-dto/order/repository.ts
import { Order } from "./order";

export interface OrderRepository {
  findPendingById(orderId: string): Order;
  save(order: Order): void;
}

export class InMemoryOrderRepository implements OrderRepository {
  private readonly orders = new Map<string, Order>();

  constructor(seed: Order[]) {
    for (const order of seed) {
      this.orders.set(order.id(), order);
    }
  }

  findPendingById(orderId: string): Order {
    const order = this.orders.get(orderId);
    if (!order) {
      throw new Error(`order ${orderId} was not found`);
    }

    if (order.status() !== "pending") {
      throw new Error(`order ${orderId} is not pending`);
    }

    return order;
  }

  save(order: Order): void {
    this.orders.set(order.id(), order);
  }

  mustGet(orderId: string): Order {
    const order = this.orders.get(orderId);
    if (!order) {
      throw new Error(`order ${orderId} was not found`);
    }

    return order;
  }
}

File: examples/go/order-repository-dto/order/repository.go
package order

import "fmt"

type Repository interface {
	FindPendingByID(orderID string) (Order, error)
	Save(order Order) error
}

type InMemoryRepository struct {
	orders map[string]Order
}

func NewInMemoryRepository(seed []Order) *InMemoryRepository {
	orders := make(map[string]Order, len(seed))
	for _, order := range seed {
		orders[order.ID()] = order
	}

	return &InMemoryRepository{orders: orders}
}

func (r *InMemoryRepository) FindPendingByID(orderID string) (Order, error) {
	order, ok := r.orders[orderID]
	if !ok {
		return Order{}, fmt.Errorf("order %s was not found", orderID)
	}

	if order.Status() != "pending" {
		return Order{}, fmt.Errorf("order %s is not pending", orderID)
	}

	return order, nil
}

func (r *InMemoryRepository) Save(order Order) error {
	r.orders[order.ID()] = order
	return nil
}

func (r *InMemoryRepository) MustGet(orderID string) Order {
	return r.orders[orderID]
}
